### ISSUE TITLE
No python version marker on importlib-metadata

### ISSUE THREAD

**Issue by oprypin** [mkdocs/mkdocs#2958]
### Discussed in https://github.com/mkdocs/mkdocs/discussions/2956

<div type='discussions-op-text'>

<sup>Originally posted by **pawamoy** August 31, 2022</sup>
The `importlib-metadata` dependency was bumped to `>=4.3` in https://github.com/mkdocs/mkdocs/commit/ec9b583b02376b909c10889fef780cdf2f25a23d.

We're now getting resolution conflicts with Flake8 because they set an upper bound on `importlib-metadata`: `<4.3`. From their answers in different issues and PRs (starting point here: https://github.com/PyCQA/flake8/issues/1503), it seems there has been quite some friction regarding this, and they likely won't change their upper bound.

Back to MkDocs: shouldn't the dependency have a marker like `python_version < '3.8'`?

Actual conflict we get (PDM output, URLs stripped):

```
Conflicts detected: 
  importlib-metadata>=4.3 (from <Candidate mkdocs 1.3.1)
  importlib-metadata; python_version < "3.8" (from <Candidate flake8-comprehensions 3.10.0)
  importlib-metadata; python_version < "3.8" (from <Candidate flake8-tidy-imports 4.8.0)
  importlib-metadata>=0.12; python_version < "3.8" (from <Candidate pytest 7.1.2)
  importlib-metadata>=3.6.0; python_version < "3.10" (from <Candidate pytest-randomly 3.12.0)
  importlib-metadata; python_version < "3.8" (from <Candidate wps-light 0.16.1)
  importlib-metadata<4.3,>=1.1.0; python_version < "3.8" (from <Candidate flake8 5.0.4)
```

We're using Python 3.8 so I believe adding the marker would help in our case at least. Unless the right marker to use is `python_version < '3.10'`.</div>

**Comment by oprypin** [comment]
> shouldn't the dependency have a marker like python_version < '3.8'?

I would welcome a pull request that makes it actually so. Right now I think the import is unconditional. I think at the moment I don't have any more context than you do

**Comment by oprypin** [comment]
btw https://github.com/pygments/pygments/pull/2155/files can be a useful example. though we shouldn't go quite as far with compatibility (no pkg_resources)

**Comment by lyz-code** [comment]
Thank you for fixing it so fast, when can we expect it to be released?

**Comment by oprypin** [comment]
I plan to develop a lot of features and then release 1.4 in ~two months. Or do you think this deserves a patch release by itself?

**Comment by lyz-code** [comment]
Yes please, I'm unable to update my dependencies:

```
-------------------------
- Updating dependencies -
-------------------------
pdm update --no-sync
Inside an active virtualenv /tmp/tmp.vcUGbo3VoE/env, reuse it.
🔒 Lock failed
Unable to find a resolution for importlib-metadata
because of the following conflicts:
  importlib-metadata; python_version < "3.8" (from <Candidate click 8.1.3 from
https://pypi.org/simple/click/>)
  importlib-metadata; python_version < "3.8" (from <Candidate flake8-comprehensions 3.10.0 from
https://pypi.org/simple/flake8-comprehensions/>)
  importlib-metadata<4.3; python_version < "3.8" (from <Candidate flake8 4.0.1 from
https://pypi.org/simple/flake8/>)
  importlib-metadata>=0.9 (from <Candidate flake8-typing-imports 1.0.0 from
https://pypi.org/simple/flake8-typing-imports/>)
  importlib-metadata>=0.9; python_version < "3.8" (from <Candidate flake8-simplify 0.19.3 from
https://pypi.org/simple/flake8-simplify/>)
  importlib-metadata>=1.0; python_version < "3.8" (from <Candidate flakeheaven 3.0.0 from
https://pypi.org/simple/flakeheaven/>)
  importlib-metadata>=4.3 (from <Candidate mkdocs 1.3.1 from https://pypi.org/simple/mkdocs/>)
To fix this, you could loosen the dependency version constraints in pyproject.toml. See
https://pdm.fming.dev/latest/usage/dependency//#solve-the-locking-failure for more details.
See /tmp/pdm-lock-lr5nhc0j.log for detailed debug log.
[ResolutionImpossible]: Unable to find a resolution
Add '-v' to see the detailed traceback
make: *** [Makefile:15: update] Error 1
```

**Comment by oprypin** [comment]
When did flake8 do this "importlib-metadata<4.3" thing?

**Comment by lyz-code** [comment]
Looks like a long time ago, I am able to bypass it using:

```toml
[tool.pdm.overrides]

"importlib-metadata" = ">=3.10"
```

Which I don't understand why it works, that's why I wanted to remove it.

If there are good reasons not to make a small release, I'll wait with the ugly patch till then

**Comment by oprypin** [comment]
Yes, almost a year ago, it seems.
https://github.com/PyCQA/flake8/commit/975a3f45334861ebd8960c00e881443c23654bca
Strange.

Anyway I would suggest to only run flake8 on a newer Python version, maybe you don't need to run it on Python 3.7 or below?  I think the problem will pop up only for old Python versions, according to that commit. Otherwise the conflict will be gone.

You could somehow set up your requirements in a way that you don't need to resolve all dependencies for all Python versions. Maybe pdm is limiting your ability to do that, I'm not familiar with it.

In MkDocs with basic tooling (tox) I don't resolve flake8 as a dependency, but just install it only for the lint step only on one Python version:

https://github.com/mkdocs/mkdocs/blob/3896a82ebf32a23cb63f53b1e84285ff7ddb2333/.github/workflows/tox.yml#L63

https://github.com/mkdocs/mkdocs/blob/3896a82ebf32a23cb63f53b1e84285ff7ddb2333/.github/workflows/tox.yml#L79

https://github.com/mkdocs/mkdocs/blob/3896a82ebf32a23cb63f53b1e84285ff7ddb2333/tox.ini#L30


Or actually, I'm not sure why merging this PR would even resolve the conflict in general, it resolves it only for Python 3.10 but I assume that pdm still needs a resolution for all versions. Has anyone ensured that pdm actually no longer complains about dependencies with this PR merged?

Or actually, maybe you can just specify in your project:
```toml
requires-python = ">=3.8"
```
and that will be the end of your problems?

**Comment by lyz-code** [comment]
Thanks for taking the time to help me debug the issue, `requires-python = ">=3.8"` or using `tox` works, so no rush.
